sqlbase: Allow the cascader to be reused within the same statement.

BramGruneir opened the pull request:
If more than one row is being cascaded, the cascader was not reusable.

For example, we have this setup:
```sql
CREATE TABLE a (
  id INT PRIMARY KEY
);
CREATE TABLE b (
  id INT PRIMARY KEY
 ,a_id INT REFERENCES a ON DELETE CASCADE
)
```

And then we insert a bunch of values into tables a and b.

So far, all the tests were doing a single delete, as in
`DELETE FROM a WHERE id = 1`

But `DELETE FROM a WHERE id > 0` would fail. Or even `DELETE FROM a`.

This fixes that issue.

The root cause was the fact that the cascader was being reused but the row
containers used for tracking deletes and updates were closed preventing further
use. The panic in #21563 was caused by trying to close the already closed row
container. But if there was another key to cascade a delete, it would have been
disasterous as well.

Fixes #21563.

Release note (sql change): Fixed a bug in cascading foreign reference actions
such that which more than one cascading operation can occur per statement. As
in, cascading from multiple deletes now works.

BramGruneir committed:
sqlbase: Allow the cascader to be reused within the same statement.

If more than one row is being cascaded, the cascader was not reusable.

For example, we have this setup:
```sql
CREATE TABLE a (
  id INT PRIMARY KEY
);
CREATE TABLE b (
  id INT PRIMARY KEY
 ,a_id INT REFERENCES a ON DELETE CASCADE
)
```

And then we insert a bunch of values into tables a and b.

So far, all the tests were doing a single delete, as in
`DELETE FROM a WHERE id = 1`

But `DELETE FROM a WHERE id > 0` would fail. Or even `DELETE FROM a`.

This fixes that issue.

The root cause was the fact that the cascader was being reused but the row
containers used for tracking deletes and updates were closed preventing further
use. The panic in #21563 was caused by trying to close the already closed row
container. But if there was another key to cascade a delete, it would have been
disasterous as well.

Fixes #21563.

Release note (sql change): Fixed a bug in cascading foreign reference actions
such that which more than one cascading operation can occur per statement. As
in, cascading from multiple deletes now works.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21799)
<!-- Reviewable:end -->

justinj commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/cascader.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-L434569IiIUM44BfFFo:-L434569IiIUM44BfFFp:bp31hdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a35ae8ecd6a943e3278b64322e71d6157a852e1/pkg/sql/sqlbase/cascader.go#L185)):*
> ```Go
> 		container.Clear(ctx)
> 	}
> 	c.deletedRows = make(map[ID]*RowContainer)
> ```

These probably both shouldn't be `deletedRows`, right? There should be an assignment for each of `deletedRows`, `originalRows`, and `updatedRows`?

---

*[pkg/sql/sqlbase/cascader.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-L434TWvtbkyxvb-ouaK:-L434TWwStyU9r10r07Z:b7w3qio) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a35ae8ecd6a943e3278b64322e71d6157a852e1/pkg/sql/sqlbase/cascader.go#L908)):*
> ```Go
> 	traceKV bool,
> ) error {
> 	defer c.clear(ctx)
> ```

Does the cascader still need to get closed at some point, or is it sufficient for it to end up having been cleared?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/sqlbase/cascader.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-L434569IiIUM44BfFFo:-L44RvN56RuTDZoANbcW:b-2rnajx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a35ae8ecd6a943e3278b64322e71d6157a852e1/pkg/sql/sqlbase/cascader.go#L185)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

These probably both shouldn't be `deletedRows`, right? There should be an assignment for each of `deletedRows`, `originalRows`, and `updatedRows`?
</blockquote></details>

Oh, this was just an old line from before I switched to a clear.  I removed both of the `make` statements.

---

*[pkg/sql/sqlbase/cascader.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-L434TWvtbkyxvb-ouaK:-L44RlCnV-4rRDor7Pdk:b-g5226s) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a35ae8ecd6a943e3278b64322e71d6157a852e1/pkg/sql/sqlbase/cascader.go#L908)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Does the cascader still need to get closed at some point, or is it sufficient for it to end up having been cleared?
</blockquote></details>

It is sufficient for it to be cleared.  This was confirmed by @knz.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/cascader.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-L434TWvtbkyxvb-ouaK:-L45xPMVwFUqtU24GyP2:b-squfvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a35ae8ecd6a943e3278b64322e71d6157a852e1/pkg/sql/sqlbase/cascader.go#L908)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

It is sufficient for it to be cleared.  This was confirmed by @knz.
</blockquote></details>

Well I just said it is sufficient for the row containers. I am making no statement about any other member of the cascader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/cascader.go, line 908 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-L434TWvtbkyxvb-ouaK:-L46drVq6M82oqc3YxBR:b3j49sj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a35ae8ecd6a943e3278b64322e71d6157a852e1/pkg/sql/sqlbase/cascader.go#L908)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Well I just said it is sufficient for the row containers. I am making no statement about any other member of the cascader.
</blockquote></details>

All the other members should be alright.  But I'll do some measurements of memory on a heavy cascading workload to ensure that that is the case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799)*
<!-- Sent from Reviewable.io -->

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21799#-:-L47TTubvuHpioBPR7P6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
TFTR!
